apis/core.oam.dev/v1alpha2/applicationrevision_types.go

```diff
@@ -40,11 +41,23 @@ type ApplicationRevisionSpec struct {
 	ScopeDefinitions map[string]ScopeDefinition `json:"scopeDefinitions,omitempty"`
 
 	// Components records the rendered components from Application, it will contains the whole K8s CR of workload in it.
-	Components []*Component `json:"components"`
+	Components map[string]RawExtensionWrapper `json:"components,omitempty"`
 
 	// ApplicationConfiguration records the rendered applicationConfiguration from Application,
 	// it will contains the whole K8s CR of trait and the reference component in it.
-	ApplicationConfiguration ApplicationConfiguration `json:"applicationConfiguration"`
+	// +kubebuilder:validation:EmbeddedResource
+	// +kubebuilder:pruning:PreserveUnknownFields
+	ApplicationConfiguration runtime.RawExtension `json:"applicationConfiguration"`
+}
+
+// RawExtensionWrapper is used to work around embedding runtime.RawExtensions.
+// If we use map[string]runtime.RawExtention or []runtime.RawExtention, the fields won't be preserved.
+// We have to add some annotations to the OpenAPISchema to preserve the fields.
+type RawExtensionWrapper struct {
```

Comment:
can we not just use the runtime.RawExtension?

Reply:
No. Unfortunately We can't...

Reply:
Changed components to be map to get rid of the wrapper